adapter/internal/sourcewatcher/sourcewatcher.go

```diff
@@ -0,0 +1,286 @@
+		Auth: gitAuth,
+	})
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error while listing remote: %v", err)
+	}
+
+	ref, err := localRepository.Head()
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error while retrieving Head: %v", err)
+	}
+
+	refName := ref.Name()
+
+	localRepositoryHash := ref.Hash().String()
+
+	for _, r := range remoteList {
+		if r.Name() == refName && !r.Hash().IsZero() && localRepositoryHash != r.Hash().String() {
+			loggers.LoggerAPI.Info("Fetching commit with hash: ", r.Hash().String(), " from remote repository")
+			pullChanges(localRepository)
+
+			err := processArtifactChanges()
+			if err != nil {
+				loggers.LoggerAPI.Errorf("Error while processing artifact changes: %v", err)
+			}
+
+			//Redeploy changes
+			artifactsMap, err = api.ProcessMountedAPIProjects()
+			if err != nil {
+				loggers.LoggerAPI.Errorf("Local api artifacts processing has failed: %v", err)
+				return
+			}
+		}
+	}
+
+}
+
+// pullChanges pulls changes from the given repository
+func pullChanges(localRepository *git.Repository){
+	conf, err := config.ReadConfigs()
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error reading configs: %v", err)
+		return
+	}
+
+	branch := conf.Adapter.SourceControl.Repository.Branch
+
+	workTree, err := localRepository.Worktree()
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error while retrieving the worktree: %v", err)
+	}
+
+	gitAuth, err := auth.GetGitAuth()
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error while authenticating the remote repository: %v", err)
+	}
+
+	pullOptions := &git.PullOptions{
+		Auth: gitAuth,
+	}
+
+	if branch != "" {
+		pullOptions.ReferenceName = plumbing.ReferenceName("refs/heads/" + branch)
+	}
+
+	err = workTree.Pull(pullOptions)
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error while pulling changes from repository: %v", err)
+	}
+}
+
+// processArtifactChanges undeploy the APIs whose artifacts are not present in the repository
+func processArtifactChanges() (err error){
+	conf, err := config.ReadConfigs()
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error reading configs: %v", err)
+		return err
+	}
+
+	apisDirName := filepath.FromSlash(conf.Adapter.SourceControl.ArtifactsDirectory + "/" + apisArtifactDir)
+	files, err := ioutil.ReadDir(apisDirName)
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Error while reading api artifacts during startup: %v", err)
+	}
+
+	// Assign the API artifacts to a temporary map
+	removedArtifacts := artifactsMap
+
+	for _, apiProjectFile := range files {
+		if apiProjectFile.IsDir() {
+			// If the artifact is present in the artifacts directory, remove it from the temporary map
+			if _, ok := removedArtifacts[apiProjectFile.Name()]; ok {
+				delete(removedArtifacts, apiProjectFile.Name())
+			}
+			continue
+		} else if !strings.HasSuffix(apiProjectFile.Name(), zipExt) {
+			continue
+		}
+		if _, ok := removedArtifacts[apiProjectFile.Name()]; ok {
+			delete(removedArtifacts, apiProjectFile.Name())
+		}
```

Comment:
When does this section get hit?

Reply:
It is hit if the path of the API project exists in the artifacts directory after pulling